src/DGmethods/examples/ex_001_periodic_advection.jl

```diff
@@ -0,0 +1,563 @@
+# we need a function that given $\vec{x}$ sets $q$.
+#
+# The initial condition is set by the solver through a function which takes $q$
+# as an `MVector` to initialize based on the pointwise coordinate location
+# `(x_1, x_2, x_3)`.
+#
+# Note: The initial condition will always be called as though the dimensionality
+# of the problem is 3. For the domain used below `x_3 = 0` when the problem is
+# actually two-dimensional; since when $x_3 = 0$ the function $\exp(\sin(2\pi
+# x_{3})) = 1$ we can safely assume the dimensionality is always $3$ in our
+# implication of the initial condition.
+function initialcondition!(Q, x_1, x_2, x_3)
+  @inbounds Q[1] = exp(sin(2π * x_1)) * exp(sin(2π * x_2)) * exp(sin(2π * x_3))
+end
+#md nothing # hide
+
+#------------------------------------------------------------------------------
+
+# ### Exact Solution
+# For periodic constant velocity advection the exact solution is trivial to
+# compute. Assuming that $\phi(x)$ is the periodically replicated initial
+# condition, the analytic solution is
+# ```math
+# q(\vec{x}, t) = \phi(\vec{x} - \vec{u} t).
+# ```
+# This will be useful later since it will allow us to check our work by
+# computing error in our solution and estimate the convergence rate.
+#
+# For a general initial condition on the unit domain the following function can
+# be used:
+function exactsolution!(dim, Q, t, x_1, x_2, x_3)
+  @inbounds begin
+    DFloat = eltype(Q)
+
+    ## trace back the point (x_1, x_2, x_3) in the velocity field and
+    ## determine where in our "original" [0, L_1] X [0, L_2] X [0, L_3] domain
+    ## this point is located
+    y_1 = mod(x_1 - DFloat(uvec[1]) * t, 1)
+    y_2 = mod(x_2 - DFloat(uvec[2]) * t, 1)
+
+    ## if we are really just 2-D we do not want to change the x_3 coordinate
+    y_3 = dim == 3 ? mod(x_3 - DFloat(uvec[3]) * t, 1) : x_3
+
+    initialcondition!(Q, y_1, y_2, y_3)
+  end
+end
+#md nothing # hide
+# The input argument `dim` is the "real" dimensionality of the problem and is
+# needed in case `uvec[3] != 0`.
+
+#------------------------------------------------------------------------------
+
+# ### Initialize the DG Method
+# We are now at the point that we can initialize the structure for the DG
+# method.  For convenience we define a function that initializes the DG method
+# over a given MPI communicator `mpicomm`, for a given `polynomialorder`, using
+# `dim` dimensions. The mesh used will be `Ne` x `Ne` elements when `dim == 2`
+# and `Ne` x `Ne` x `Ne` elements when `dim == 3`. The floating point type of
+# the computation is `DFloat` and whether the CPU or GPU is used is determined
+# by `ArrayType`; `ArrayType === Array` is for the CPU and `ArrayType ===
+# CuArray` is for NVIDIA GPUs.
+#
+# Note: This whole code chunk is in a function block
+function setupDG(mpicomm, dim, Ne, polynomialorder, DFloat=Float64,
+                 ArrayType=Array)
+
+  @assert ArrayType === Array
```

Comment:
Hit this line when trying to run
```julia
spatialdiscretization = setupDG(mpicomm, dim, Ne, polynomialorder, Float64, CuArray)
```
so I'm guessing you can't run this example on a GPU just yet.

Reply:
That is right.  GPU support is coming.